sporkfish/endgame_tablebase.py

```diff
@@ -121,26 +184,33 @@ def query(self, board: Board) -> Optional[chess.Move]:
         :rtype: Optional[chess.Move]
         """
 
-        try:
-            if self._db:
-                if not self._should_query(board):
-                    return None
-
-                # We need to use a chess.Board() to be compatible with the endgame tablebase
-                # This is a small performace hit but is miniscule compared to searching
-                cboard = chess.Board()
-                cboard.set_fen(board.fen())
-                for move in cboard.legal_moves:
-                    cboard.push(move)
-                    # WDL Score: returns 2 if the side to move is winning, 0 if the position is a draw and -2 if the side to move is losing.
-                    # Returns 1 in case of a cursed win and -1 in case of a blessed loss.
-                    # Mate can be forced but the position can be drawn due to the fifty-move rule.
-                    # We check WDL score < 0 here, since we pushed our legal move and so its the opponents turn.
-                    # If they are losing, we are winning.
-                    if self._db.probe_wdl(cboard) < 0:
-                        return move
-                    cboard.pop()
-                return None
-            return None
-        except chess.syzygy.MissingTableError:
-            return None
+        if self._db and self._should_query(board):
+            # We need to use a chess.Board() to be compatible with the endgame tablebase
+            # This is a small performace hit but is miniscule compared to searching
+            cboard = chess.Board()
+            cboard.set_fen(board.fen())
+            best_move = None
+            best_dtz = -sys.maxsize
+            best_category = EndgameTablebase.DTZCategory.UNCONDITIONAL_LOSS
+            for move in board.legal_moves:
+                cboard.push(move)
+                # Refer to https://python-chess.readthedocs.io/en/latest/syzygy.html
+                # Invert the dtz as we are looking from the perspective of the opponent
+                dtz = -opp_dtz if (opp_dtz := self._db.get_dtz(cboard)) else None
```

Comment:
Good catch, actually the behaviour is slightly weird. I thought if your board pieces matches the Szygzy base name then we guaranteed is guaranteed, but from the tests it didn't seem to be the case. Should double check.

Reply:
Ahhh I think I get your point now....sorry

The decision on how to choose the `best_dtz` must be seen from 'our perspective'. As mentioned above, I am happy to leave it as is.

```
        # We want to save our cursed loss as quickly as possible
        if category == EndgameTablebase.DTZCategory.CURSED_LOSS:
            return max(dtz, best_dtz)
        # We want make the unconditional loss last as long as possible, in case they run out of time
        elif category == EndgameTablebase.DTZCategory.UNCONDITIONAL_LOSS:
            return min(dtz, best_dtz)
        elif category == EndgameTablebase.DTZCategory.UNCONDITIONAL_DRAW:
            return 0
        # We want to unconditionally win as quickly as possible
        elif category == EndgameTablebase.DTZCategory.UNCONDITIONAL_WIN:
            return min(dtz, best_dtz)
        # We want to extend our cursed win as much as possible, in case they run out of time
        elif category == EndgameTablebase.DTZCategory.CURSED_WIN:
            return max(dtz, best_dtz)
```

Reply:
Sorry - just making sure I get you - you mean -100 <= n <= -1 (I understand this one as unconditional loss) or do you mean blessed loss?

I think the other 2 points make sense. B) actually will still be a problem and will lead to a draw. I was thinking to reinstate @michelle-lai-swl original solution (pick 1 from all the best dtz score) but actually would prefer to see a case where this happens deterministically (even though I suppose it is deterministic because of seeding) so we can reverse engineer the dynamics of the rounding. I tried to follow the thread and python-chess code but got nowhere.

Reply:
Sorry I meant unconditional loss